File: BeckersRixen.m
function [Ma,U,S,V]=BeckersRixen(M0,nmin,nmax,perc,tol)
%This is a Beckers and Rixen 2003 implementation of the iterated
%EOF gap-filling technique.  The method takes a gappy dataset in M0
%M0 and produces an analysis dataset Ma of the same size.
%M0 should have rows as time-series and columns as datagrid patterns.
%
%Each iteration proceeds by SVD of the previous Ma,
% [Un,Sn,Vn]=svds(Ma(old),num)
% Then, by truncation to fewer modes n<num, an update is made
% Ma(new)=Un*Sn*Vn'
% This continues until Sn converges.
%
% The algorithm begins with n=1 and num=n+1, runs to convergence
% based on reducing error to tol*(initial error), and estimates final error.
% Next, the iteration is repeated with higher n and error is estimated
% When the estimated error increases, the previous n result is saved.
% nmax sets an upper limit of the number of modes to include.
%
% The error is estimated by repeating perc percent of the data with
% zeros introduced.  The synthetic zeros occur in the same proportion
% by row of M0 as they do in the original.

nnow=1;
sm=size(M0);
if ~exist('nmin')  % This is the default % of modes to use
    nmin=1;
end
if ~exist('nmax')  % This is the default % of modes to use
    nmax=round(min(sm)./2);
end
if ~exist('perc')  % This is the default % of matrix to use as error est.
    perc=10;
end
if ~exist('tol')  % This is the default error tolerance
    tol=1e-4;
end
nmax=min(min(sm)-1,nmax);

%Count the zeros by row
zcount=sum(M0'==0)./sm(2);

% Add on the error estimate rows
added=ceil(sm(2)*perc/100);   % how many to add?
Mp=zeros(sm(1),sm(2)+added);  % a bigger matrix with the added rows
Mp(:,1:sm(2))=M0;
Md=Mp;
ndx=randi(sm(2),added,1);   % make a random guess of columns to include

for ii=1:added
  while prod(M0(:,ndx(ii)))==0
      ndx(ii)=randi(sm(2),1,1);  %try again until you find a column without gaps
  end
  Mp(:,sm(2)+ii)=M0(:,ndx(ii));  % Duplicate the original nonzero data
  Md(:,sm(2)+ii)=Mp(:,sm(2)+ii).*(rand(1,sm(1))>zcount)'; %zeroing some
end

zdx=find(Md==0);
adx=(sm(2)+1):(sm(2)+added);

initerror=sum(sum((Mp(:,adx)-Md(:,adx)).^2));  % The initial error, now lets reduce it!
itererror=initerror;
nerror(nnow)=0;
decerror=true;

% Heres the loop for the number of modes used
for nnow=1:nmax
 itererror=initerror;
 errorup=2*initerror;
 olderrorup=3*initerror;
 Mdnew=Md;
 
% pause
 while abs(errorup./olderrorup-1)>tol%&&errorup>tol*initerror
   
   Md(zdx)=Mdnew(zdx);
     
   [U,S,V]=svds(Md,nnow+1);
   
   for jj=(nnow+1):length(S)
    S(jj,jj)=0;   % Truncate the modes
    U(:,jj)=0;   % Truncate the modes
    V(:,jj)=0;   % Truncate the modes
   end
   Mdnew=U*S*V';  % Form the new guess
   
   itererror=sum(sum((Mp(:,adx)-Mdnew(:,adx)).^2)); %Find new error
   olderror=sum(sum((Mp(:,adx)-Md(:,adx)).^2)); %Find error update
   
   olderrorup=errorup;
   errorup=sum(sum((Mp(:,adx)-Md(:,adx)).^2))-itererror; %Find error update
   [abs(errorup./olderrorup-1),tol];
   
   %figure(100)
   %clf
   %plot(Mp(zdx),Mp(zdx),'-',Md(zdx),Mp(zdx),'k.',Mdnew(zdx),Mp(zdx),'r.')
   %legend('line','Md','Mdnew','Location','northwest')
   %axis equal
   %drawnow
   
 end
 nerror(nnow)=itererror; % Check to ensure error is decreasing with increasing n
 if nerror(nnow)>nerror(max(1,nnow-1))
  break
 end
 nnow=nnow+1;  % Take more nodes
end

Ma=Md(:,1:sm(2));

File: BeckersRixen.py
import numpy as np

def BeckersRixen(M0, test, nmin=1, nmax=None, val_perc=10, tol=1e-4):
    nnow = 1
    sm = M0.shape
    m, n = sm
    test_m, test_n = test.shape
    assert m == test_m
    
    if nmax is None:  # Default number of modes to use
        nmax = round(np.min(sm) / 2)
        
    nmax = min(np.min(sm) - 1, nmax)
    
    # Count the zeros by row, want to match ratio of zeros in test set to ensure it works well
    zcount = np.sum(test == 0, axis=1) / test_n
    
    # Add on the error estimate rows
    val_n = int(np.ceil(n * val_perc / 100))   # How many to train on?
    Mp = M0.copy()
    Md = Mp.copy()
    ndx = np.random.randint(n, size=val_n)   # Make a random guess of columns to include
    
    for ii in range(val_n):
        give_up = 100
        while np.prod(M0[:, ndx[ii]]) == 0 and give_up > 0:
            ndx[ii] = np.random.randint(n)  # Try again until you find a column without gaps (is this necessary?)
            give_up -= 1
        Md[:, ndx[ii]] *= (np.random.rand(m) > zcount)  # Zeroing some

    Md = np.concatenate((Md, test), axis = 1) # Append test set so it's also affected by the algorithm

    zdx = np.where(Md == 0)
    
    initerror = np.sum((Mp[:, ndx] - Md[:, ndx]) ** 2)  # The initial error, now let's reduce it!
    itererror = initerror
    nerror = np.zeros(nmax)
    
    # Here's the loop for the number of modes used
    # Don't reset the places at zdx to 0, BeckersRixen paper says not to
    while nnow <= nmax:
        itererror = initerror
        errorup = 2 * initerror
        olderrorup = 3 * initerror
        Mdnew = Md.copy()
        
        while np.abs(errorup / olderrorup - 1) > tol:
            Md[zdx] = Mdnew[zdx]
            U, S, V = np.linalg.svd(Md, full_matrices=False)
            
            Mdnew = np.dot(U[:,:nnow + 1], np.dot(np.diag(S[:nnow + 1]), V[:nnow + 1]))  # Form the new guess
            
            itererror = np.sum((Mp[:, ndx] - Mdnew[:, ndx]) ** 2)  # Find new error
            olderror = np.sum((Mp[:, ndx] - Md[:, ndx]) ** 2)  # Find error update
            
            olderrorup = errorup
            errorup = np.sum((Mp[:, ndx] - Md[:, ndx]) ** 2) - itererror  # Find error update
            
        nerror[nnow-1] = itererror  # Check to ensure error is decreasing with increasing n
        if nnow > 1 and nerror[nnow-1] > nerror[nnow-2]:
            break
        
        nnow += 1  # Take more nodes
    
    Ma = Md
    
    return Ma